pennylane/transforms/get_unitary_matrix.py

```diff
@@ -0,0 +1,141 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A transform to obtain the matrix representation of a quantum circuit.
+"""
+from functools import wraps
+import numpy as np
+from pennylane.wires import Wires
+import pennylane as qml
+
+
+def get_unitary_matrix(circuit, wire_order=None):
+    r"""Construct the matrix representation of a quantum circuit.
+
+    Args:
+        circuit (.QNode, .QuantumTape, or Callable): A quantum node, tape, or function that applies quantum operations.
+        wire_order (Sequence[Any], optional): Order of the wires in the quantum circuit. Defaults to the order in which the wires appear in the quantum function.
+
+    Returns:
+         function: Function which accepts the same arguments as the QNode or quantum function.
+         When called, this function will return the unitary matrix.
+
+    **Example**
+
+    Consider the following function (the same applies for a QNode or tape):
+
+    .. code-block:: python3
+
+        def circuit(theta):
+            qml.RX(theta, wires=1)
+            qml.PauliZ(wires=0)
+
+
+    We can use ``get_unitary_matrix`` to generate a new function
+    that returns the unitary matrix corresponding to the function ``circuit``:
+
+
+    >>> get_matrix = get_unitary_matrix(circuit)
+    >>> theta = np.pi/4
+    >>> get_matrix(theta)
+    array([[ 0.92387953+0.j,  0.+0.j ,  0.-0.38268343j,  0.+0.j],
+       [ 0.+0.j,  -0.92387953+0.j,  0.+0.j,  0. +0.38268343j],
+       [ 0. -0.38268343j,  0.+0.j,  0.92387953+0.j,  0.+0.j],
+       [ 0.+0.j,  0.+0.38268343j,  0.+0.j,  -0.92387953+0.j]])
+
+
+    Note that since ``wire_order`` was not specified, the default order ``[1, 0]``
+    for ``circuit`` was used, and the unitary matrix corresponds to the operation
+    :math:`Z\otimes R_X(\theta)`. To obtain the matrix for :math:`R_X(\theta)\otimes Z`,
+    specify ``wire_order=[0, 1`` in the function call:
+
+    >>> get_matrix = get_unitary_matrix(circuit, wire_order=[0, 1])
+
+    You can also get the unitary matrix for operations on a subspace of a
+    larger Hilbert space. For example, with the same function ``circuit`` and
+    ``wire_order=["a", 0, "b", 1]`` you obtain the :math:`16\times 16` matrix for
+    the operation :math:`I\otimes Z\otimes I\otimes  R_X(\theta)`.
+    """
+
+    wires = wire_order
+
+    @wraps(circuit)
+    def wrapper(*args, **kwargs):
+
+        if isinstance(circuit, qml.QNode):
+            # user passed a QNode, get the tape
+            circuit.construct(args, kwargs)
+            tape = circuit.qtape
+
+            if wires is None:  # if no wire ordering is specified, take wire list from tape
+                wire_order = tape.wires
+            else:
+                wire_order = Wires(wires)
+
+        elif isinstance(circuit, qml.tape.QuantumTape):
+            # user passed a tape
+            tape = circuit
+            if wires is None:
```

Comment:
Re. the pylint "too many branches" complaint, since this code block appears in all three cases

```
if wires is None:
    wire_order = tape.wires
else:
    wire_order = Wires(wires)
```

perhaps moving it out to after the instances checks will appease it?

Reply:
Good idea 🙂 You could even reduce it to a single line:

```python
wire_order = tape.wires if wires is None else Wires(wires)
```